### PR TITLE
[risk=low][RW-8126] Show a link to the signed DUCC in the Profile

### PR DESCRIPTION
**Pull request by jmthibault79**
If a user has signed **the latest version of the** DUCC, a link to view it will appear in their profile:
signed-ducc

If a user has not signed the DUCC, they will not see this link in their profile.  If they happen to type `/signed-ducc` manually into their browser (unlikely!) they would see this:
no-ducc

Users who have signed older versions of the DUCC, they will also not see this link in their profile.  If they happen to type `/signed-ducc` manually into their browser (unlikely!) they would see this:
old-ducc

 See also the prototype at #6651 

---
**PR checklist**

- [x] This PR meets the Acceptance Criteria in the JIRA story
- [x] The JIRA story has been moved to Dev Review
- [x] This PR includes appropriate unit tests
- [x] I have added explanatory comments where the logic is not obvious
- [x] I have run and tested this change locally, and my testing process is described here
- [x] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later
- [ ] If this includes an API change, I have run the E2E tests on this change against my local server with [yarn test-local](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples) because this PR won't be covered by the CircleCI tests 
- [x] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P)
- [x] If this change impacts deployment safety (e.g. removing/altering APIs which are in use) I have documented these in the description
- [x] If this includes an API change, I have updated the appropriate Swagger definitions and updated the appropriate API consumers
  * AoU UI
  * [Perf tests](https://github.com/broadinstitute/mcnulty/blob/develop/src/test/scala/services/AoU.scala)
  * [Researcher Directory export](https://github.com/all-of-us/workbench/wiki/Researcher-Directory-(RDR-export))
  * Cron tasks - for Offline*Controllers
  * SumoLogic - for EgressAlert

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/access/UserAccessModuleMapper.java [hunk 1]
```diff
@@ -1,36 +1,20 @@
...
 import org.pmiops.workbench.db.model.DbUserAccessModule;
 import org.pmiops.workbench.model.AccessModuleStatus;
 import org.pmiops.workbench.utils.mappers.CommonMappers;
 import org.pmiops.workbench.utils.mappers.MapStructConfig;
 
 @Mapper(
     config = MapStructConfig.class,
-    uses = {CommonMappers.class, AccessModuleNameMapper.class},
-    nullValueMappingStrategy = NullValueMappingStrategy.RETURN_DEFAULT)
+    uses = {CommonMappers.class, AccessModuleNameMapper.class})
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/access/UserAccessModuleMapper.java:13]: unrelated simplification - I can take this out of the PR if you like

File: ui/src/app/pages/profile/data-user-code-of-conduct.tsx [hunk 2]
```diff
@@ -66,10 +70,6 @@ const styles = reactStyles({
   },
 });
 
-// TODO parameterize these
```
Comment (jmthibault79) [inline ui/src/app/pages/profile/data-user-code-of-conduct.tsx:69]: parameterized, now in utils/code-of-conduct

File: ui/src/app/pages/profile/data-user-code-of-conduct.tsx [hunk 3]
```diff
@@ -112,10 +130,10 @@ interface ReadOnlyProps extends DuccTextInputProps {
   signatureState: DuccSignatureState;
 }
 const ReadOnlyTextField = (props: ReadOnlyProps) =>
-  props.signatureState === DuccSignatureState.UNSIGNED ? (
-    <DuccTextInput {...fp.omit(['signatureState'], props)} disabled={true} />
-  ) : (
+  props.signatureState === DuccSignatureState.SIGNED ? (
```
Comment (jmthibault79) [inline ui/src/app/pages/profile/data-user-code-of-conduct.tsx:133]: changes up to L150 are just logic sign flips

File: ui/src/app/pages/profile/demographic-survey-panel.tsx [hunk 4]
```diff
@@ -14,10 +13,10 @@ interface Props {
 export const DemographicSurveyPanel = (props: Props) => {
   const { demographicSurveyCompletionTime, firstSignInTime, onClick } = props;
   return (
-    <div style={{ marginTop: '1rem', marginLeft: '1rem' }}>
+    <div style={styles.panel}>
```
Comment (jmthibault79) [inline ui/src/app/pages/profile/demographic-survey-panel.tsx:16]: common style to the Demo Survey and DUCC panels

File: ui/src/app/pages/profile/data-user-code-of-conduct.tsx [hunk 5]
```diff
@@ -80,6 +80,19 @@ export const enum DuccSignatureState {
   SIGNED,
 }
 
+const VersionFailure = () => (
+  <>
+    <div>
+      Error: 'we are unable to display this version of the Data User Code of
```
Comment (jmthibault79) [inline ui/src/app/pages/profile/data-user-code-of-conduct.tsx:86]: covers both cases: unsigned DUCC, or version is older than the first one we formatted for signed display (4).

File: ui/src/app/pages/profile/data-user-code-of-conduct.tsx [hunk 6]
```diff
@@ -144,43 +156,52 @@ const InitialsAgreement = (props: InitialsProps) => (
...
-          Proceed
-        </Button>
-      </FlexRow>
-    )}
-  </>
-);
+const DuccContentPage = (props: ContentProps) => {
+  const versionInfo = getDuccRenderingInfo(props.versionToRender);
+
+  return versionInfo ? (
```
Comment (jmthibault79) [inline ui/src/app/pages/profile/data-user-code-of-conduct.tsx:167]: recommended to review with whitespace diffs off

File: api/src/main/java/org/pmiops/workbench/config/WorkbenchConfigMapper.java [hunk 7]
```diff
@@ -77,5 +77,6 @@ default void mapRuntimeImages(WorkbenchConfig source, @MappingTarget ConfigRespo
   @Mapping(target = "rasClientId", source = "config.ras.clientId")
   @Mapping(target = "rasLogoutUrl", source = "config.ras.logoutUrl")
   @Mapping(target = "freeTierBillingAccountId", source = "config.billing.accountId")
+  @Mapping(target = "currentDuccVersions", source = "config.access.currentDuccVersions")
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/config/WorkbenchConfigMapper.java:80]: This is the list of DUCC versions which are acceptable for RT/CT membership.  They are NOT the versions which can be shown in SIGNED mode.

File: ui/src/app/components/data-user-code-of-conduct.tsx [hunk 8]
```diff
@@ -0,0 +1,552 @@
...
+  CONTENT,
+  SIGNATURE,
+}
+
+export const enum DuccSignatureState {
+  UNSIGNED,
+  SIGNED,
+}
+
+const SignedVersionFailure = (props: { duccSignedVersion: number }) => {
```
Comment (jmthibault79) [inline ui/src/app/components/data-user-code-of-conduct.tsx:83]: opinions welcome for this error text

File: ui/src/app/pages/profile/data-user-code-of-conduct.tsx [hunk 9]
```diff
@@ -144,43 +160,55 @@ const InitialsAgreement = (props: InitialsProps) => (
...
-          disabled={props.buttonDisabled}
-          onClick={() => props.onClick()}
-        >
-          Proceed
-        </Button>
-      </FlexRow>
-    )}
-  </>
-);
+const DuccContentPage = (props: ContentProps) => {
```
Comment (jmthibault79) [inline ui/src/app/pages/profile/data-user-code-of-conduct.tsx:168]: probably want to review this with whitespace diffs off

**Review by evrii — APPROVED** [review]
Looks good to me. If the user has signed a ducc, and it has expired, what would we expect to see on the /signed-ducc page?

It currently looks like this:
image
Not sure if we want to show different or not.

File: api/src/main/java/org/pmiops/workbench/access/UserAccessModuleMapper.java (hunk 1 above)
Comment (jmthibault79) [reply api/src/main/java/org/pmiops/workbench/access/UserAccessModuleMapper.java:13]: I didn't see any default being used ... usually better not to use one IMO, if avoidable.

File: ui/src/app/routing/signed-in-app-routing.tsx [hunk 10]
```diff
@@ -297,10 +297,18 @@ export const SignedInRoutes = () => {
...
           }}
-          // see RW-8126 and RW-8296
           signatureState={DuccSignatureState.UNSIGNED}
         />
       </AppRoute>
+      <AppRoute exact path='/signed-ducc'>
+        <DataUserCodeOfConductPage
+          routeData={{
+            title: 'Data User Code of Conduct (Signed)',
+            minimizeChrome: true,
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply ui/src/app/routing/signed-in-app-routing.tsx:307]: the name also confuses me

File: ui/src/app/pages/profile/data-user-code-of-conduct.spec.tsx [hunk 11]
```diff
@@ -147,4 +164,38 @@ describe('DataUserCodeOfConduct', () => {
...
+    const wrapper = component(DuccSignatureState.SIGNED);
+    expect(
+      wrapper.find('[data-test-id="ducc-content-page"]').exists()
+    ).toBeFalsy();
+    expect(
+      wrapper.find('[data-test-id="ducc-signature-page"]').exists()
+    ).toBeFalsy();
+  });
+
+  it('should not display the Content and Signature pages in SIGNED mode if the user has signed an older DUCC', async () => {
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply ui/src/app/pages/profile/data-user-code-of-conduct.spec.tsx:190]: Same data-test-ids.

This is testing the logic of showContentPage / showSignaturePage

**Comment by jmthibault79** [comment]
> If the user has signed a ducc, and it has expired, what would we expect to see on the /signed-ducc page?
I would expect that the user wouldn't be able to access it, because they'd be redirected to AAR instead.  But I'll try that out...

**Comment by jmthibault79** [comment]
@evrii good catch regarding the expired DUCC!  Fixed now by adding the routing guard.

**Pull request merged**